core/src/main/java/com/linecorp/armeria/server/throttling/FileBasedDynamicThrottlingStrategy.java

```diff
@@ -0,0 +1,158 @@
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.base.Ascii;
+import com.google.common.base.Splitter;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.primitives.Floats;
+import com.sun.nio.file.SensitivityWatchEventModifier;
+
+import com.linecorp.armeria.common.Request;
+import com.linecorp.armeria.common.RpcRequest;
+import com.linecorp.armeria.server.ServiceRequestContext;
+
+/**
+ * A {@link ThrottlingStrategy} to read and maintain a map from service API name to {@link ThrottlingStrategy}.
+ * The {@link FileBasedDynamicThrottlingStrategy} will monitor a json mapping file for updates, whose format is
+ * keys for every service API name with values is a string representation of {@link ThrottlingStrategy}.
+ */
+public class FileBasedDynamicThrottlingStrategy<T extends Request> extends ThrottlingStrategy<T> {
+    private static final Logger logger = LoggerFactory.getLogger(FileBasedDynamicThrottlingStrategy.class);
+    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+    private static final Splitter COLON_SPLITTER = Splitter.on(':').trimResults();
+
+    private final WatchService watchService;
+    private final Path settingFile;
+    private final Thread watchThread;
+    private volatile Map<String, ThrottlingStrategy<T>> strategies = ImmutableMap.of();
+
+    /**
+     * Creates a new instance.
+     */
+    public FileBasedDynamicThrottlingStrategy(String settingFile) {
+        this.settingFile = FileSystems.getDefault().getPath(settingFile);
+        try {
+            watchService = FileSystems.getDefault().newWatchService();
+            this.settingFile.getParent().register(watchService,
+                                                  new WatchEvent.Kind[] {
+                                                          StandardWatchEventKinds.ENTRY_CREATE,
+                                                          StandardWatchEventKinds.ENTRY_MODIFY
+                                                  },
+                                                  SensitivityWatchEventModifier.HIGH);
+        } catch (IOException e) {
+            throw new IllegalStateException("Could not start watch service.", e);
+        }
+        updateThrottlingSettings();
+        watchThread = startWatching();
+    }
+
+    private Thread startWatching() {
+        Thread thread = new Thread(() -> {
+            try {
+                while (true) {
+                    try {
+                        WatchKey key = watchService.take();
+                        for (WatchEvent<?> event : key.pollEvents()) {
+                            // We only register "ENTRY_MODIFY" so the context is always a Path.
+                            final Path changed = (Path) event.context();
+                            if (changed.equals(settingFile.getFileName())) {
```

Comment:
This comparison doesn't work when symlinks are involved (e.g., a Kubernetes config map where the specified human-readable filename like `settings.txt` will be a symlink to a timestamped uuid type filename like `settings.fq38rua3r908auf.txt`, and the event gets triggered with the timestamped file, not symlink file update.

I'm not sure if it's the best solution, but resolving the real paths during this check could help
https://github.com/curioswitch/curiostack/commit/0164c45c76df6897ed7bc5c7ebc2efe6da938272

Reply:
Nice catch! And thanks for reference.
To focus on other basic throttoring, I removed FileBasedDynamicThrottlingStrategy.java from this PR  strategies. Let me try dig more by myself on next PR.